### PR TITLE
use as_dict() because label_matcher_dict excludes unit

### PR DESCRIPTION
**Pull request by Abuelodelanada**
## Issue

This PR fixes https://github.com/canonical/grafana-agent-operator/issues/61

## Solution

We stop using `label_matcher_dict` because removes `unit`.

```diff
@@ -700,10 +700,17 @@ class COSAgentRequirer(Object):
 
                 # Apply labels to the scrape jobs
                 for static_config in job.get("static_configs", []):
+                    topo_as_dict = topology.as_dict(excluded_keys=["charm_name"])
                     static_config["labels"] = {
                         # Be sure to keep labels from static_config
                         **static_config.get("labels", {}),
-                        **topology.as_dict(excluded_keys=["charm_name"]),
+                        # TODO: We should add a new method in juju_topology.py
+                        # that like `as_dict` method, returns the keys with juju_ prefix
+                        **{
+                            "juju_{}".format(key): value
+                            for key, value in topo_as_dict.items()
+                            if value
+                        },
```

## TODO:

Add itest mimicking [this deployment](https://github.com/canonical/grafana-agent-operator/issues/61#issuecomment-1977366788)

### CODE REVIEW

File: lib/charms/grafana_agent/v0/cos_agent.py [hunk 1]
```diff
@@ -700,10 +700,17 @@ def metrics_jobs(self) -> List[Dict]:
 
                 # Apply labels to the scrape jobs
                 for static_config in job.get("static_configs", []):
+                    topo_as_dict = topology.as_dict(excluded_keys=["charm_name"])
                     static_config["labels"] = {
                         # Be sure to keep labels from static_config
                         **static_config.get("labels", {}),
-                        **topology.label_matcher_dict,
+                        # TODO: We should add a new method in juju_topology.py
+                        # that like `as_dict` method, returns the keys with juju_ prefix
```
Comment (sed-i) [inline lib/charms/grafana_agent/v0/cos_agent.py:708]: Open an issue in cos-lib?
Or if you consider it in-scope then have a tandem PR?

File: tests/integration/test_juju_info_cos_agent.py [hunk 2]
```diff
@@ -12,6 +12,7 @@
 from pytest_operator.plugin import OpsTest
 
 agent = SimpleNamespace(name="agent")
+hwo = SimpleNamespace(charm="hardware-observer", name="hwo")
```
Comment (sed-i) [inline tests/integration/test_juju_info_cos_agent.py:15]: `hwo` added but no new assertions. Is that intentional?

**Review by sed-i — APPROVED** [review]

File: tests/integration/test_juju_info_cos_agent.py (hunk 2 above)
Comment (Abuelodelanada) [reply tests/integration/test_juju_info_cos_agent.py:15]: We are checking the existence of missing `juju_unit` label here:

```python
    for targets in unit_targets:
        for target in targets:
            target_labels = target["labels"].keys()
            assert topology_labels.issubset(target_labels)
```

Do you think we can be more specific? 🤔

File: lib/charms/grafana_agent/v0/cos_agent.py (hunk 1 above)
Comment (Abuelodelanada) [reply lib/charms/grafana_agent/v0/cos_agent.py:708]: I've opened an issue for this here: https://github.com/canonical/cos-lib/issues/18

I would open 2 in-tandem PR here and there

Comment (lucabello) [inline lib/charms/grafana_agent/v0/cos_agent.py:708]: **nitpick**: I'd suggest maybe adding a link to the issue you created as an extra comment line

**Review by lucabello — APPROVED** [review]

**Pull request merged**